Online deletion docs

mDuo13 opened the pull request:
This PR includes the following changes (links to the staged versions, which require Ripple VPN):

- Adds [a new concept doc about online deletion](https://developers-staging.ripple.com/online-deletion.html) to replace [the dead wiki page](https://web.archive.org/web/20150422095042/https://wiki.ripple.com/Online_deletion)
- Wraps the online deletion and history sharding docs into a new subcategory called [Ledger History](https://developers-staging.ripple.com/ledger-history.html). (I moved some of the content from an earlier version of this PR into the "Ledger History" page.)
- Adds 3 new tutorials around online deletion:
    - [Changing the amount of history you store](https://developers-staging.ripple.com/configure-online-deletion.html)
    - [Using advisory delete with cron for scheduled deletions](https://developers-staging.ripple.com/configure-advisory-deletion.html)
    - [Storing full history](https://developers-staging.ripple.com/configure-full-history.html)
- Updates the Capacity Planning doc with [a table of how much disk space is required to store different amounts of history](https://developers-staging.ripple.com/capacity-planning.html#historical-data).

The `.uxf` files are source files for the diagrams, so they don't need to be directly reviewed.

mDuo13 committed:
Online deletion: concept 1st draft w/ diagrams

mDuo13 committed:
Advisory delete, full history tutorials. History disk space estimates.

mDuo13 committed:
Online deletion: simpler tutorial

mDuo13 committed:
Move storage disk requirements table to capacity planning doc

mDuo13 commented on content/tutorials/manage-the-rippled-server/configuration/configure-full-history.md:
```diff
@@ -0,0 +1,98 @@
+
+        [import_db]
+      	type=RocksDB
+      	path=/tmp/full_history_dump/
```
The syntax for the `[import_db]` stanza was not clearly specified in the config file so I took an educated guess. Please confirm.

jhaaaa replied:
I know this is just an example and not sure if this matters - but maybe the `type` should be set to `NuDB`, just to reinforce our recommendation that you use NuDB for full history?

jhaaaa replied:
Thank you for writing this tutorial, Rome - great info, so concise, and I think the topic is of great interest.

mDuo13 replied:
Good catch on the RocksDB thing. I changed the `[node_db]` stanza but missed the later `[import_db]` example on my first pass. I think technically they _can_ be different types, but most likely if you're importing from another DB that has full history, it'll be NuDB→NuDB.

mtrippled commented on content/concepts/the-rippled-server/online-deletion.md:
```diff
@@ -0,0 +1,126 @@
+# Online Deletion
+[[Source]<br/>](https://github.com/ripple/rippled/blob/master/src/ripple/app/misc/SHAMapStoreImp.cpp "Source")
+
+The online deletion feature lets the `rippled` server delete the server's local copy of old ledger versions to keep disk usage from rapidly growing over time. Online deletion happens automatically by default, but can be configured to run only when prompted. [New in: rippled 0.27.0][]
```
Online deletion is not on by default, it needs to be configured using the online_delete parameter in the [node_db] section. Then it's automatic unless configured to run only when triggered.

ximinez replied:
Does "by default" mean "in the default / sample config file"?

mDuo13 replied:
That was the intention. Rewording to make that clearer.

mtrippled commented on content/concepts/the-rippled-server/online-deletion.md:
```diff
@@ -0,0 +1,126 @@
+
+## Background
+
+The `rippled` server stores data from recent ledger versions in its _ledger store_. This data accumulates as the XRP Ledger network creates new ledger versions and validates them through the [consensus process](intro-to-consensus.html).
```
Do we call this "ledger store" or "node store"? (no opinion here either way but I think comments in source code as well as object names tend to refer to it as "node store".

mDuo13 replied:
The documentation calls it the "ledger store" because the previous term "node store" (still common in the code, etc.) has shown a history of being confusing.

mDuo13 committed:
Online deletion: more interlinking

mtrippled commented on content/concepts/the-rippled-server/online-deletion.md:
```diff
@@ -0,0 +1,126 @@
+
+### Fetching History
+
+When it starts, a `rippled` server's first priority is to get a complete copy of the latest validated ledger. From there, it keeps up with advances in the ledger progress. If configured to do so, the server also backfills ledger history up to a configured amount, which must be equal or less than the cutoff beyond which online deletion is configured to delete.
```
Default ledger_history is 256. It can be set as low as 0 (no ledger history to fetch). However, online_delete has a minimum of 256. online_delete must be equal to or greater than ledger_history. The only reason to configure ledger_history explicitly with also online_delete configured, is to fetch more history upon first initializing rippled. Otherwise, the server will always attempt to maintain at minimum the value configured for online_delete.

mDuo13 replied:
You mean, the server backfills gaps in the ledger history up to the configured online_delete setting even if it's higher than the ledger_history setting? (I'm making a distinction between "gaps" and ledgers missing before the earliest ledger locally available, here.)

mDuo13 replied:
Clarification (from offline convo): this is only true if advisory_delete is on (as mentioned in the other comment below)

mtrippled commented on content/concepts/the-rippled-server/online-deletion.md:
```diff
@@ -0,0 +1,126 @@
+
+When it starts, a `rippled` server's first priority is to get a complete copy of the latest validated ledger. From there, it keeps up with advances in the ledger progress. If configured to do so, the server also backfills ledger history up to a configured amount, which must be equal or less than the cutoff beyond which online deletion is configured to delete.
+
+The server can backfill history from before it became synced, as well as filling in any gaps in the history it has collected after syncing. (Gaps in ledger history can occur if a server temporarily becomes too busy to keep up with the network, loses its network connection, or suffers other temporary issues.) To backfill history, the server requests data from its peer `rippled` servers. The amount ...
```
Actually, let me clarify the relationship further between online_delete, advisory_delete, and ledger_history.
If advisory_delete is not configured, then rippled will only seek to maintain ledger_history ledgers between online_delete intervals: if there is a gap earlier than what online_delete has been configured for, then rippled will not attempt to fill the gap.

If advisory_delete is configured, then rippled will attempt to ensure that gaps are filled between online_delete intervals.

mtrippled commented on content/concepts/the-rippled-server/online-deletion.md:
```diff
@@ -0,0 +1,126 @@
+- The server may store up to twice the configured number of ledger versions if online deletion is set to run automatically. (Each time it runs, it reduces the number of stored ledger versions to the configured number.)
+- If advisory delete is enabled, the server stores all the ledger versions that it has acquired and built until someone calls the [can_delete method][].
+
+    For example, if you call `can_delete` with a value of `now` once per day and an `online_delete` value of 2000, the server typically stores 2000 ledger versions at minimum and just over a full day's worth of ledger versions at maximum.
```
The server will store between 1 and 2 full days of ledgers in this case, assuming that >=2000 ledgers happen each day.

mtrippled commented on content/concepts/the-rippled-server/online-deletion.md:
```diff
@@ -0,0 +1,126 @@
+
+With online deletion enabled and running automatically (that is, with advisory delete disabled), the total amount of ledger data stored should remain at minimum equal to the number of ledger versions the server is configured to keep, with the maximum being roughly twice that many.
+
+When online deletion runs, it does not reduce the size of the database files on disk; it only makes space within those files available to be reused for new data. To actually reduce the disk space allocated to the database files, you must restart `rippled`.
```
Running manual commands to vauum the SQLite databases is the only way to reduce the data size of sqlite databases. Node store size should reduce (rocksdb or nudb), but SQLITE won't shrink. Let's not advise that people manually reduce SQLite data files, it's time-consuming and risk prone if done incorrectly.

mDuo13 replied:
Gotcha, that's an outdated thing from the old wiki page.

mtrippled commented on content/concepts/the-rippled-server/online-deletion.md:
```diff
@@ -0,0 +1,126 @@
+
+If you stop the server or it crashes while online deletion is running, online deletion resumes after the server is restarted and the server becomes fully synced.
+
+To temporarily disable online deletion, you can use the [can_delete method][] with an argument of `never`. This change persists until the server restarts or you re-enable online deletion by calling [can_delete][can_delete method] again. For more information on controlling when online deletion happens, see [Advisory Deletion](#advisory-deletion).
```
"can_delete never" persists indefinitely, between restarts, until can_delete sets another value.

mtrippled commented on content/concepts/the-rippled-server/online-deletion.md:
```diff
@@ -0,0 +1,126 @@
+
+    **Caution:** If you run `rippled` with `online_delete` disabled, then later enable `online_delete` and restart the server, the server disregards but does not delete existing ledger history that your server already downloaded while `online_delete` was disabled. To save disk space, delete your existing history before re-starting the server with `online_delete` back on.
+
+- **`[ledger_history]`** - Specify a number of validated ledgers, equal to or less than `online_delete`. If the server does not have at least this many validated ledger versions, it attempts to backfill them by fetching the data from peers.
```
Actually, let me clarify the relationship further between online_delete, advisory_delete, and ledger_history.
If advisory_delete is not configured, then rippled will only seek to maintain ledger_history ledgers between online_delete intervals: if there is a gap earlier than what online_delete has been configured for, then rippled will not attempt to fill the gap.

If advisory_delete is configured, then rippled will attempt to ensure that gaps are filled between online_delete intervals.

mtrippled commented on content/concepts/the-rippled-server/online-deletion.md:
```diff
@@ -0,0 +1,126 @@
+
+    The following diagram shows the relationship between `online_delete` and `ledger_history` settings:
+
+    ![Ledgers older than `online_delete` are automatically deleted. Ledgers newer than `ledger_history` are backfilled. Ledgers in between are kept if available but not backfilled](img/online_delete-vs-ledger_history.png)
```
Actually, let me clarify the relationship further between online_delete, advisory_delete, and ledger_history.
If advisory_delete is not configured, then rippled will only seek to maintain ledger_history ledgers between online_delete intervals: if there is a gap earlier than what online_delete has been configured for, then rippled will not attempt to fill the gap.

If advisory_delete is configured, then rippled will attempt to ensure that gaps are filled between online_delete intervals.

mtrippled commented on content/concepts/the-rippled-server/online-deletion.md:
```diff
@@ -0,0 +1,126 @@
+
+### Advisory Deletion
+
+By default, online deletion happens automatically and periodically. If the `advisory_delete` setting is enabled, online deletion only happens when an administrator triggers it using the [can_delete method][]. You can run this command with a scheduled job to trigger automatic deletion based on clock time instead of the number of ledger versions closed.
```
If online_delete is configured, it happens automatically. If not configured (the default), then it doesn't. advisory_delete has no meaning if it's not configured.

And can_delete can be run manually to maintain as much history is desired, and to delete only those earlier ledgers that you don't want any more.

mDuo13 replied:
Reworded to make this more precise.

mtrippled commented on content/concepts/the-rippled-server/online-deletion.md:
```diff
@@ -0,0 +1,126 @@
+
+You can also use advisory delete for other reasons, such as if you want to manually confirm that the transaction data is backed up to a separate server before deleting it.
+
+The `can_delete` API method can enable advisory deletion (with the value `never`) or disable advisory delete (with the value `always`). These setting changes persist until you restart the `rippled` server, at which point the settings in the config file override them.
```
They're persisted.
never: don't do online_delete
always: do online_delete as soon as the online_delete number of ledgers has been retained.
now: do online_delete at the next ledger close or as soon as the online_delete number of ledgers have been acquired, whichever is greater.
<integer> or <ledger hash>: Delete up to and including that ledger as soon as at least the online_delete number of ledgers have been acquired.

^^
What I just described is documented at: https://developers.ripple.com/can_delete.html

mtrippled commented on content/tutorials/manage-the-rippled-server/configuration/configure-advisory-deletion.md:
```diff
@@ -0,0 +1,96 @@
+# Configure Advisory Deletion
+
+In its default configuration, the `rippled` server automatically deletes outdated history of XRP Ledger state and transactions as new ledger versions become available. If your server uses most of its hardware resources during peak hours, you can configure the server to delete ledgers only when prompted by a command scheduled to run during off-peak hours to reduce the chances of online deletion
```
By "default" rippled does not do online_delete. It needs to be configured. However, the example config file shipped right now with the source code happens to configure online_delete.

mtrippled commented on content/tutorials/manage-the-rippled-server/configuration/configure-advisory-deletion.md:
```diff
@@ -0,0 +1,96 @@
+      	online_delete=2000
+      	advisory_delete=1
+
+    - Set `advisory_delete` to `1` to require approval before running online deletion. (Set it to `0` to run online deletion automatically as new ledger versions become available.)
```
require the can_delete API call to trigger (not "approval" from anybody). Set to '0' to run online deletion once the configured number of ledgers are acquired.

mtrippled commented on content/tutorials/manage-the-rippled-server/configuration/configure-advisory-deletion.md:
```diff
@@ -0,0 +1,96 @@
+        $ rippled --conf=/etc/opt/ripple/rippled.cfg can_delete now
+
+    The response indicates the maximum ledger index that may the server may delete from its ledger store. For example, the following message indicates that ledger versions up to and including ledger index 43633667 are to be deleted:
+
```
are to be deleted once the latest ledger acquired is >= that value plus the value of online_delete.

mDuo13 replied:
Added a note following the code example to clarify this.

ximinez replied:
This sentence reads awkwardly to me. Instead of "are to be deleted", I think it should say "are available to be deleted", "may be deleted", or something like that.

jhaaaa replied:
suggested edit:

>maximum ledger index that may the server may delete from its ledger store.

to

>maximum ledger index that the server may delete from its ledger store.

mtrippled commented on content/tutorials/manage-the-rippled-server/configuration/configure-advisory-deletion.md:
```diff
@@ -0,0 +1,96 @@
+
+        5 1 * * * rippled --conf /etc/opt/ripple/rippled.cfg can_delete now
+
+    Be sure that you schedule the command to run based on the time zone your server's clock is configured to use. Ripple recommends running all production servers with the UTC time zone.
```
Why tell people what time zone to use?

mtrippled commented on content/tutorials/manage-the-rippled-server/configuration/configure-advisory-deletion.md:
```diff
@@ -0,0 +1,96 @@
+
+    Be sure that you schedule the command to run based on the time zone your server's clock is configured to use. Ripple recommends running all production servers with the UTC time zone.
+
+    **Tip:** You do not need to schedule a `cron` job to run online deletion if you have `advisory_delete` disabled. In that case, `rippled` runs online deletion automatically when the server's oldest available ledger is twice as old as the number of ledgers to keep after deletion.
```
Actually, it's triggered once the difference between the earliest ledger acquired and the latest level acquired are at least the value of online_delete. This also means that, eventually, the oldest available ledger will be roughly twice that value.

mtrippled commented on content/tutorials/manage-the-rippled-server/configuration/configure-full-history.md:
```diff
@@ -0,0 +1,98 @@
+      	#online_delete=2000
+      	#advisory_delete=0
+
+    You can store full history with either RocksDB or NuDB as the key-value store behind the ledger store. NuDB requires less RAM but RocksDB requires less disk space. For more information, see [Capacity Planning](capacity-planning.html).
```
Don't store full history with rocksdb. Just nudb. Based on RAM utilization of each. RAM requirements of rocksdb for full history are unknown.

mtrippled commented on content/tutorials/manage-the-rippled-server/configuration/configure-full-history.md:
```diff
@@ -0,0 +1,98 @@
+
+    You can store full history with either RocksDB or NuDB as the key-value store behind the ledger store. NuDB requires less RAM but RocksDB requires less disk space. For more information, see [Capacity Planning](capacity-planning.html).
+
+    If you are using RocksDB, add `compression=1` to enable Snappy compression, which reduces the disk space necessary to store full history (at a cost of greater CPU usage when reading or writing history).
```
It may also reduce CPU utilization because less time spent in I/O. Compression is enabled by default. There's no reason to not compress rocksdb unless somebody wants to test this. Recommend leaving it alone, also recommend not even mentioning it.

mDuo13 committed:
Online deletion revisions

- Split ledger history section into a parent concept page w/ online
deletion and history sharding as child pages
- Clarified how much data is stored in general
- Covered the special case for backfilling with advisory_delete enabled
(including updating the diagram)
- Changed full history recommendation to NuDB only

mDuo13 committed:
Capacity planning: unify storage sections

mDuo13 committed:
Add 'empty' case to server_info/state ref

mDuo13 commented on content/tutorials/manage-the-rippled-server/installation/capacity-planning.md:
```diff
@@ -116,32 +117,49 @@ For best performance in enterprise production environments, Ripple recommends ru
+
+| Real Time Amount | Number of Ledger Versions | Disk Space Required (RocksDB) | Disk Space Required (NuDB) |
+|:-----------------|:--------------------------|:------------------------------|:--|
+| 2 hours          | 2,000                     | ***(TBD)*** | ***(TBD)*** |
```
I'm going to test these and update later.

mDuo13 committed:
Capacity planning: add 2-hour estimates

mDuo13 committed:
State DB error, how to fix, & relation to key-value store type

mDuo13 committed:
Fix broken disk storage links

ximinez commented on content/tutorials/manage-the-rippled-server/configuration/configure-advisory-deletion.md:
```diff
@@ -0,0 +1,98 @@
+# Configure Advisory Deletion
+
+The default config file sets `rippled` to automatically delete outdated history of XRP Ledger state and transactions as new ledger versions become available. If your server uses most of its hardware resources during peak hours, you can configure the server to delete ledgers only when prompted by a command scheduled to run during off-peak hours to reduce the chances of online deletion.
```
"to run during off-peak hours to reduce the chances of online deletion"... doing what? I assume it's something like "running during peak hours", but the sentence reads awkwardly to me.

intelliot replied:
Maybe “to run during off-peak hours to mitigate the performance impact of online deletion.”

mDuo13 replied:
Good catch. This was a copy-edit failure. Changed to:

> If your server uses most of its hardware resources during peak hours, you can configure the server to delete ledgers only when prompted by a command scheduled to run during off-peak hours, so that online deletion is less likely to impact server performance.

ximinez commented on content/concepts/the-rippled-server/ledger-history/online-deletion.md:
```diff
@@ -0,0 +1,107 @@
+
+If you stop the server or it crashes while online deletion is running, online deletion resumes after the server is restarted and the server becomes fully synced.
+
+To temporarily disable online deletion, you can use the [can_delete method][] with an argument of `never`. This change persists until you re-enable online deletion by calling [can_delete][can_delete method] again, even if you change the . For more information on controlling when online deletion happens, see [Advisory Deletion](#advisory-deletion).
```
"even if you change the " what?

mtrippled replied:
Yeah, not clearly expressed here, whatever it is.

mDuo13 replied:
Good catch. Changed to:

> This change persists until you re-enable online deletion by calling [can_delete] again.

_Edit: Actually, disabling advisory deletion in the config file is kind of redundant with this anyway, so simplified again._

ximinez commented on content/concepts/the-rippled-server/ledger-history/online-deletion.md:
```diff
@@ -0,0 +1,107 @@
+
+    The default config file specifies 2000 for this value. This cannot be less than 256, because some events like [Fee Voting](fee-voting.html) and the [Amendment Process](amendments.html#amendment-process) update only every 256 ledgers.
+
+    **Caution:** If you run `rippled` with `online_delete` disabled, then later enable `online_delete` and restart the server, the server disregards but does not delete existing ledger history that your server already downloaded while `online_delete` was disabled. To save disk space, delete your existing history before re-starting the server with `online_delete` back on.
```
The converse is also true. Should that be mentioned here?

mDuo13 replied:
Good point. Changed the last sentence to:

> To save disk space, delete your existing history before re-starting the server after changing the `online_delete` setting.

ximinez commented on content/concepts/the-rippled-server/ledger-history/online-deletion.md:
```diff
@@ -0,0 +1,107 @@
+
+You can use advisory deletion with a scheduled job to trigger automatic deletion based on clock time instead of the number of ledger versions closed. If your server is heavily used, the extra load from online deletion can cause your server to fall behind and temporarily de-sync from the consensus network. If this is the case, you can use advisory deletion and schedule online deletion to happen on ...
+
+Alternatively, you can use advisory deletion for other reasons, such as if you want to manually confirm that transaction data is backed up to a separate server before deleting it.
```
Another reason is wanting to manually confirm that data is processed by another task before deleting it. Is that worth mentioning?

mDuo13 replied:
Good point. Rephrased:

> You can use advisory deletion for other reasons. For example, you may want to manually confirm that transaction data is backed up to a separate server before deleting it. Alternatively, you may want to manually confirm that a separate task has finished processing transaction data before you delete that data.

ximinez reviewed (changes requested):
A few suggestions and questions.

thejohnfreeman commented on content/concepts/the-rippled-server/ledger-history/online-deletion.md:
```diff
@@ -0,0 +1,107 @@
+    If you call [can_delete][can_delete method] on a regular basis, the maximum amount of data the server stores is either of the following, whichever is larger:
+
+    - A number of ledger versions approximately equal to twice the `online_delete` value. (After deletion, this is reduced to approximately the `online_delete` value.)
+    - Ledger versions spanning an amount of time that is approximately twice the interval between `can_delete` calls. (After deletion, this is reduced to approximately one interval's worth of data.)
```
If I understand this section correctly, l'd like to add a hint before each case, like this:

- If you call `can_delete` too frequently, a number of ledger versions approximately equal to twice the `online_delete` value. (After deletion, this is reduced to approximately the `online_delete` value.)
- If you call `can_delete` too infrequently, ledger versions spanning an amount of time that is approximately twice the interval between `can_delete` calls. (After deletion, this is reduced to approximately one interval's worth of data.)

mtrippled replied:
If it's called too frequently (meaning before online_delete quantity of ledgers have passed), then can_delete has no effect. Rotation won't happen until at least the minimum configured online_delete ledgers have passed.

thejohnfreeman replied:
I understand that the subsequent calls will (likely) not trigger a delete, but I get the impression that this section of documentation isn't about whether an individual call has an effect, it is about the upper bound on data stored as a function of call frequency. I would like a rule of thumb, and the way it's phrased is not very intuitive to me. Feel free to rephrase "too frequently" to "very frequently" or "frequently enough" or similar; it means the same to me. I'm trying to picture two extremes on a spectrum of `can_delete` frequency, and assign storage bounds to those extremes, to get a sense of the relationship between call frequency and storage cost.

mDuo13 replied:
Rewrote this section with an example for each case:

> The amount of data the server stores depends on how often you call [can_delete][can_delete method] and the how big an interval of time your `online_delete` setting represents:
>
>    - If you call `can_delete` _more often_ than your `online_delete` interval, the server stores at most a number of ledger versions approximately equal to **twice the `online_delete` value**. (After deletion, this is reduced to approximately the `online_delete` value.)
>
>        For example, if you call `can_delete` with a value of `now` once per day and an `online_delete` value of 50,000, the server typically stores up to 100,000 ledger versions before running deletion. After running deletion, the server keeps at least 50,000 ledger versions (about two days' worth). With this configuration, approximately every other `can_delete` call results in no change because the server does not have enough ledger versions to delete.
>
>    - If you call `can_delete` _less often_ than your `online_delete` interval, the server stores at most ledger versions spanning an amount of time that is approximately **twice the interval between `can_delete` calls**. (After deletion, this is reduced to approximately one interval's worth of data.)
>
>        For example, if you call `can_delete` with a value of `now` once per day and an `online_delete` value of 2000, the server typically stores up to two full days' worth of ledger versions before running deletion. After running deletion, the server keeps approximately one day's worth (about 25,000 ledger versions), but never fewer than 2000 ledger versions.

thejohnfreeman commented on content/concepts/the-rippled-server/ledger-history/online-deletion.md:
```diff
@@ -0,0 +1,107 @@
+    - A number of ledger versions approximately equal to twice the `online_delete` value. (After deletion, this is reduced to approximately the `online_delete` value.)
+    - Ledger versions spanning an amount of time that is approximately twice the interval between `can_delete` calls. (After deletion, this is reduced to approximately one interval's worth of data.)
+
+    For example, if you call `can_delete` with a value of `now` once per day and an `online_delete` value of 2000, the server typically stores up two full day's worth of ledger versions before running deletion. After running deletion, the server keeps approximately 1 day's worth, but never less than 2000 ledger versions.
```
One correction: "never less than 2000" -> "never fewer than 2000"

One suggested addition: "the server keeps approximately 1 day's worth (around 20,000 ledger versions)"

jhaaaa replied:
suggested nit edit:

>the server typically stores up two full day's worth

to

>the server typically stores up to two full day's worth

thejohnfreeman commented on content/concepts/the-rippled-server/ledger-history/online-deletion.md:
```diff
@@ -0,0 +1,107 @@
+
+    **Caution:** If you run `rippled` with `online_delete` disabled, then later enable `online_delete` and restart the server, the server disregards but does not delete existing ledger history that your server already downloaded while `online_delete` was disabled. To save disk space, delete your existing history before re-starting the server with `online_delete` back on.
+
+- **`[ledger_history]`** - Specify a number of validated ledgers, equal to or less than `online_delete`. If the server does not have at least this many validated ledger versions, it attempts to backfill them by fetching the data from peers.
```
1. Why are `ledger_history` and `fetch_depth` in brackets, but `online_delete` and `advisory_delete` are not?

2. Am I right in thinking of `ledger_history` as a *minimum* number of validated ledgers to keep, which will be backfilled when necessary? If so, can we explicitly say "minimum" in its description?

mtrippled replied:
In the config file, sections are enclosed in brackets whereas sub-items within a section are simply k=v. online_delete and advisory_delete are within the section [node_db].

thejohnfreeman replied:
Is `ledger_history` a config section? Why does the documentation describe it as a number? Do you just put the number on its own line under `[ledger_history]`? If so, I cannot wait for the TOML migration...

mDuo13 replied:
Yes, `[ledger_history]` and `[fetch_depth]` are sections, and yes, you put the number on its own line under `[ledger_history]`.

And don't count on the config file changing soon. Backwards-compatibility and all.

jhaaaa commented on content/concepts/the-rippled-server/ledger-history/ledger-history.md:
```diff
@@ -0,0 +1,56 @@
+
+The [consensus process](intro-to-consensus.html) creates a chain of [validated ledger versions](ledgers.html), each derived from the previous one by applying a set of transactions. Every `rippled` server stores ledger versions and transaction history locally. The amount of transaction history a server stores depends on how long that server has been online and how much history it is configured to  ...
+
+Servers in the peer-to-peer XRP Ledger network share transactions and other data with each other as part of the consensus process. Each server each independently builds each new ledger version and compares results with its trusted validators to ensure consistency. (If a consensus of trusted validators disagrees with a server's results, that server fetches the necessary data from its peers to achi ...
```
Suggested nit edit - if it doesn't change your intended meaning:

>Each server each independently builds each new ledger version

to

>Each server independently builds each new ledger version

jhaaaa commented on content/concepts/the-rippled-server/ledger-history/ledger-history.md:
```diff
@@ -0,0 +1,56 @@
+
+## Fetching History
+
+When it starts, a `rippled` server's first priority is to get a complete copy of the latest validated ledger. From there, it keeps up with advances in the ledger progress. If configured to do so, the server also backfills ledger history up to a configured amount, which must be equal or less than the cutoff beyond which online deletion is configured to delete.
```
suggested nit edit:

>which must be equal or less than the cutoff

to 

>which must be equal to or less than the cutoff

jhaaaa commented on content/concepts/the-rippled-server/ledger-history/ledger-history.md:
```diff
@@ -0,0 +1,56 @@
+
+**Tip:** Unlike some cryptocurrency networks, servers in the XRP Ledger do not need full history to know the current state and keep up with current transactions.
+
+For instructions setting up full history, see [Configure Full History](configure-full-history.html).
```
Suggested edit:

>For instructions setting up full history,

to

>For instructions on setting up full history,

jhaaaa commented on content/concepts/the-rippled-server/ledger-history/online-deletion.md:
```diff
@@ -0,0 +1,107 @@
+
+The `rippled` server stores [ledger history](ledger-history.html) in its _ledger store_. This data accumulates over time.
+
+Inside the ledger store, ledger data is "de-duplicated": data that doesn't change from version to version is only stored once. The records themselves in the ledger store do not indicate which ledger version(s) contain them; part of the work of online deletion is identifying which records are only used by outdated ledger versions. This process is time consuming and affects the disk I/O and applica ...
```
Nit - stray ":"? 

>ledger data is "de-duplicated": data

to

>ledger data is "de-duplicated" data

jhaaaa replied:
Is this a valid suggested edit? 🤔 

>is "de-duplicated" data that doesn't change from version to version is only stored once.

to

>is "de-duplicated" data that doesn't change from version to version and is only stored once.

thejohnfreeman replied:
@jhaaaa Your suggested edit is valid, but I actually prefer the way it was written. I like short sentences, and the colon splits one long sentence into two short ones, but this is just my opinion.

mDuo13 replied:
The colon was not stray. It indicates that the remainder is an independent clause that clarifies the preceding clause. I see where you mis-read "de-duplicated" as being an adjective that modifies the following data, but it's actually a predicate adjective that modifies the previous instance of "data".

I don't like connecting the clauses with "that" because it implies that there is another meaning or use of "de-duplicated".  To make it easier for others to avoid the same mistake, I've rephrased it into two separate sentences:

> Inside the ledger store, ledger data is "de-duplicated". In other words, data that doesn't change from version to version is only stored once.

mtrippled commented on content/concepts/the-rippled-server/ledger-history/ledger-history.md:
```diff
@@ -0,0 +1,56 @@
+
+## Fetching History
+
+When it starts, a `rippled` server's first priority is to get a complete copy of the latest validated ledger. From there, it keeps up with advances in the ledger progress. If configured to do so, the server also backfills ledger history up to a configured amount, which must be equal or less than the cutoff beyond which online deletion is configured to delete.
```
Default ledger_history is 256, but can be set as low as 0.

mDuo13 replied:
That's good to know, but I'm going to leave it out here. (It would be more appropriate in a config reference than in this context.)

mtrippled commented on content/tutorials/manage-the-rippled-server/installation/capacity-planning.md:
```diff
@@ -55,45 +55,47 @@ The example `rippled-example.cfg` file has the `type` field in the `[node_db]` s
 [NuDB](https://github.com/vinniefalco/nudb#introduction) is an append-only key-value store that is optimized for SSD drives.
 
-NuDB has nearly constant performance and memory footprints regardless of the amount of data being [stored](#storage). NuDB _requires_ a solid-state drive, but uses much less RAM than RocksDB to access a large database.
+NuDB has nearly constant performance and memory footprints regardless of the [amount of data being stored](#disk-space). NuDB _requires_ a solid-state drive, but uses much less RAM than RocksDB to access a large database.
```
SSD's are recommended for all configurations, not just nudb.

mDuo13 replied:
Yes, that is clarified a couple other places in this document, in particular lines 44-50 which lead into this section.

jhaaaa commented on content/references/rippled-api/public-rippled-methods/server-info-methods/server_info.md:
```diff
@@ -283,7 +283,7 @@ The `info` object may have some arrangement of the following fields:
 | `build_version`                     | String                    | The version number of the running `rippled` version. |
 | `closed_ledger`                     | Object                    | (May be omitted) Information on the most recently closed ledger that has not been validated by consensus. If the most recently validated ledger is available, the response omits this field and includes `validated_ledger` instead. The member fields are the same as the `validated_ledger` field. |
-| `complete_ledgers`                  | String                    | Range expression indicating the sequence numbers of the ledger versions the local rippled has in its database. This may be a disjoint sequence, for example `24900901-24900984,24901116-24901158`. |
+| `complete_ledgers`                  | String                    | Range expression indicating the sequence numbers of the ledger versions the local rippled has in its database. This may be a disjoint sequence, for example `24900901-24900984,24901116-24901158`. If the server does not have any complete ledgers (for example, it just started syncing with the network), this is the string `empty`. |
```
super nit:

>rippled

to

>`rippled`

jhaaaa commented on content/tutorials/manage-the-rippled-server/configuration/configure-advisory-deletion.md:
```diff
@@ -0,0 +1,98 @@
+
+- Your server has enough disk space to store your chosen amount of history in its ledger store.
+
+    See [Capacity Planning](capacity-planning.html) for details of how much storage is required for different configuration. With advisory deletion enabled, the maximum history a server may accumulate before deletion is equal to the number of ledger versions configured in the `online_delete` setting **plus** the amount of time between online deletion prompts.
```
suggested nit edit:

>for different configuration.

to

>for different configurations.

jhaaaa commented on content/tutorials/manage-the-rippled-server/configuration/configure-advisory-deletion.md:
```diff
@@ -0,0 +1,98 @@
+    The lowest ledger index in `complete_ledgers` should increase after online deletion.
+
+    Deletion may take several minutes to complete when it runs, depending on how busy your server is and how much history you delete at a time.
+
```
I appreciate step 5 and the tips it provides. 🙌

jhaaaa commented on content/tutorials/manage-the-rippled-server/configuration/configure-advisory-deletion.md:
```diff
@@ -0,0 +1,98 @@
+
+If online deletion does not seem to be running after configuring it, try the following:
+
+- Check that the user who configured the `cron` job has permissions to run the `rippled` server as a commandline client.
```
This troubleshooting section is great!

For this particular troubleshooting tip above - would it be useful to mention this in the step for creating the `cron` job? Or is it obvious to a user who would be performing that task? 

And just to be sure I understand why this needs to be true -- the user that created the `cron` job will be the user used to issue the `can_delete` method at the scheduled time.

mDuo13 replied:
Regarding your second point, yes, that's correct.

For the first point, I think that should be fine because the previous step has you run `rippled`, so I would expect that to come up before that point anyway. But it's in the troubleshooting in case users skipped past that.

jhaaaa commented on content/tutorials/manage-the-rippled-server/configuration/configure-online-deletion.md:
```diff
@@ -0,0 +1,75 @@
+
+1. Decide how many ledger versions' worth of history to store.
+
+    New ledger versions are usually validated 3 to 4 seconds apart, so the number of ledger versions corresponds roughly to the amount of time you want to store. See [Capacity Planning](capacity-planning.html) for details of how much storage is required for different configuration.
```
suggested nit edit:

>for different configuration.

to

>for different configurations.

jhaaaa commented on content/tutorials/manage-the-rippled-server/configuration/configure-online-deletion.md:
```diff
@@ -0,0 +1,75 @@
+
+0. Periodically check your server's `complete_ledgers` range using the [server_info method][] to confirm that ledgers are being deleted.
+
+    After online deletion runs, the `complete_ledgers` range reflects that older ledgers are no longer available. As your server accumulates history, the total number of ledgers available should slowly increase to twice the `online_delete` value you configured, then decrease when online deletion runs.
```
Great explanation of expected behavior.

mDuo13 committed:
Online deletion: edits per reviews, new diagram of how it works

jhaaaa commented on content/concepts/the-rippled-server/ledger-history/online-deletion.md:
```diff
@@ -0,0 +1,119 @@
+- The server may store up to twice the configured number of ledger versions if online deletion is set to run automatically. (Each time it runs, it reduces the number of stored ledger versions to approximately the configured number.)
+- If advisory deletion is enabled, the server stores all the ledger versions that it has acquired and built until its administrator calls the [can_delete method][].
+
+    The amount of data the server stores depends on how often you call [can_delete][can_delete method] and the how big an interval of time your `online_delete` setting represents:
```
nit: change

>and the how big an interval

to

>and how big an interval

jhaaaa reviewed (approved):
Left just one more nit of an edit. Otherwise, this looks fantastic to me. Such a huge leap forward in providing crucial details about the behavior of online deletion. Thank you!

mDuo13 committed:
Online deletion typo fix

mDuo13 commented:
Rebased on master and re-staged with the latest changes.

mtrippled commented on content/concepts/the-rippled-server/ledger-history/online-deletion.md:
```diff
@@ -0,0 +1,119 @@
+The online deletion settings configure how many ledger versions the `rippled` server should keep available in the ledger store at a time. However, the specified number is a guideline, not a hard rule:
+
+- The server never deletes data more recent than the configured number of ledger versions, but it may have less than that amount available if it has not been running for long enough or if it lost sync with the network at any time. (The server attempts to backfill at least some history; see [fetching history](ledger-history.html#fetching-history) for details.)
+- The server may store up to twice the configured number of ledger versions if online deletion is set to run automatically. (Each time it runs, it reduces the number of stored ledger versions to approximately the configured number.)
```
Will store up to twice the configured ledger versions if online deletion is not delayed. If delayed, ledgers will continue to accumulate. In fact, online deletion takes awhile to complete, in which time a handful of new ledgers have been created. So in practice, slightly more than double will be held if all is functioning properly.

mtrippled commented on content/concepts/the-rippled-server/ledger-history/online-deletion.md:
```diff
@@ -0,0 +1,119 @@
+
+    The amount of data the server stores depends on how often you call [can_delete][can_delete method] and how big an interval of time your `online_delete` setting represents:
+
+    - If you call `can_delete` _more often_ than your `online_delete` interval, the server stores at most a number of ledger versions approximately equal to **twice the `online_delete` value**. (After deletion, this is reduced to approximately the `online_delete` value.)
```
See above, it will be slightly above twice (unless online deletion is delayed).

mtrippled replied:
Though I don't know if you want to be that particular. Maybe more generic language like "in the vicinity of" rather than "up to", because it could actually go a bit higher.

mDuo13 replied:
It already says "approximately equal" here, so I think that's fine as-is. I added a more detailed explanation above (around line 25) where you also pointed it out.

mtrippled commented on content/references/rippled-api/admin-rippled-methods/logging-and-data-management-methods/can_delete.md:
```diff
@@ -47,7 +47,7 @@ The request includes the following optional parameter:
 | `Field`      | Type              | Description                               |
 |:-------------|:------------------|:------------------------------------------|
-| `can_delete` | String or Integer | The maximum ledger to allow to be deleted. For `ledger_index` or `ledger_hash`, see [Specifying Ledgers][]. `never` sets the value to 0, and effectively disables online deletion until another `can_delete` is appropriately called.  `always` sets the value to the maximum possible ledger (4294967295), and online deletion occurs as of each configured `online_delet ...
+| `can_delete` | String or Integer | The [Ledger Index][] of the maximum ledger version to allow to be deleted. The special case `never` disables online deletion. The special case `always` enables automatic online deletion as if advisory deletion was disabled. The special case `now` allows online deletion one time at the next validated ledger that meets or exceeds the configured `online_delete` v ...
```
Also maybe note that can_delete returns the ledger sequence when executed. So if somebody wants to know that value without making any changes, they can simply call can_delete with no parameters.

mDuo13 replied:
Yes, the lines before and after this table made that clear, but I've changed it to be more consistent with how these tables are styled elsewhere in the docs.

mtrippled commented on content/tutorials/manage-the-rippled-server/configuration/configure-full-history.md:
```diff
@@ -0,0 +1,92 @@
+        rm -r /var/lib/rippled/db/rocksdb/*
+        rm /var/lib/rippled/db/*.db
+        rm /var/lib/rippled/db/*.sqlite
+
```
I've seen SQLite sometimes createing files with a slightly different extension. Basically, wipe out the whole directory:
rm -fr /var/lib/rippled/db/*

mtrippled commented on content/tutorials/manage-the-rippled-server/installation/capacity-planning.md:
```diff
@@ -53,47 +53,49 @@ The example `rippled-example.cfg` file has the `type` field in the `[node_db]` s
 #### More About Using RocksDB
 
-[RocksDB](https://rocksdb.org/docs/getting-started.html) is an embeddable persistent key-value store that is optimized for rotational disks.
+[RocksDB](https://rocksdb.org/docs/getting-started.html) is an embeddable persistent key-value store that works well with both rotational disks and solid state disks.
```
No rippled works well with rotational disks!

mtrippled commented:
:+1:

mDuo13 committed:
Online deletion edits per review

- Cleaned up can_delete method fields, errors
- Changed commands to delete old db files in full history tutorial
- Clarified RocksDB w/ rotational disks
- Clarified how many ledgers may be stored w/ online deletion
